File: typescript/src/scenarios.ts
import type { FaultTarget, FaultConfig } from './fault-types.js';

export interface ChaosScenario {
  name: string;
  description: string;
  faults: readonly { target: FaultTarget; config: FaultConfig; durationMs?: number }[];
  expectedBehavior: string;
  assertions: readonly string[];
}

interface ScenarioBuilder {
  name: string;
  description: string;
  faults: { target: FaultTarget; config: FaultConfig; durationMs?: number }[];
  expectedBehavior: string;
  assertions: string[];
}

/** Build a ChaosScenario with a fluent-ish config object. */
export function defineScenario(builder: ScenarioBuilder): ChaosScenario {
  if (!builder.name) {
    throw new Error('Scenario name is required');
  }
  if (builder.faults.length === 0) {
    throw new Error('Scenario must have at least one fault');
  }
  return {
    name: builder.name,
    description: builder.description,
    faults: builder.faults,
    expectedBehavior: builder.expectedBehavior,
    assertions: builder.assertions,
  };
}

File: python/src/mcp_chaos_monkey/scenarios.py
from __future__ import annotations

from dataclasses import dataclass, field

from .fault_types import FaultConfig, FaultTarget


@dataclass
class ScenarioFault:
    target: FaultTarget
    config: FaultConfig
    duration_ms: int | None = None


@dataclass
class ChaosScenario:
    name: str
    description: str
    faults: list[ScenarioFault]
    expected_behavior: str
    assertions: list[str] = field(default_factory=list)


def define_scenario(
    *,
    name: str,
    description: str,
    faults: list[ScenarioFault],
    expected_behavior: str,
    assertions: list[str] | None = None,
) -> ChaosScenario:
    """Build a ChaosScenario with validation."""
    if not name:
        raise ValueError("Scenario name is required")
    if not faults:
        raise ValueError("Scenario must have at least one fault")
    return ChaosScenario(
        name=name,
        description=description,
        faults=faults,
        expected_behavior=expected_behavior,
        assertions=assertions or [],
    )
